### PR TITLE
runltp-ng: Recover on timeout

### PR DESCRIPTION
**Pull request by cfconrad**
Improve timeout handling. The idea is to retry the sequence of
commands once one failed in timeout.
The reboot function of the backend is used to recover the SUT.

This is a approach of how we could handle the timeout. If it goes in the
right direction
I will change all the other shell command calls.

### CODE REVIEW

File: backend.pm [hunk 1]
```diff
@@ -179,6 +179,24 @@ sub run_cmd
 	wantarray? ($ret, @log) : $ret;
 }
 
+sub run_cmds
+{
+	my ($self, $cmds,%args) = @_;
```
Comment (metan-ucw) [inline backend.pm]: Missing space after ,

File: backend.pm [hunk 2]
```diff
@@ -179,6 +179,24 @@ sub run_cmd
...
+
+	push(@log, {cmd=>'', ret=>0, log=>''}) unless (@{$cmds});
+
+	for my $cmd (@{$cmds}) {
+		my @ret = run_cmd($self, $cmd, $args{timeout});
+		my @output = @ret[1 .. $#ret];
+		push(@log, {cmd=>$cmd, ret=>$ret[0], log => \@output});
+
+		return wantarray? @log : $ret[0]
+		unless (defined($ret[0]) && $ret[0] == 0);
```
Comment (metan-ucw) [inline backend.pm]: Maybe it's just me being perl beginner, but I do find this construct a bit confusing, you can easily miss that there is no semicolon at the end of the return line and wonder what this is supposed to do.

File: backend.pm [hunk 3]
```diff
@@ -522,7 +540,9 @@ sub set_logfile
 sub check_cmd
 {
 	my ($self, $cmd) = @_;
-	return run_cmd($self, $cmd) != 127;
+	my $ret = run_cmd($self, $cmd);
+	return unless(defined($ret));
+	return $ret != 127;
```
Comment (metan-ucw) [inline backend.pm]: Have you actually managed to get undefined from the run_cmd here? I guess that it does not happen unless we set timeout, which we should do here anyway.

File: utils.pm [hunk 4]
```diff
@@ -121,86 +121,59 @@ sub collect_sysinfo
 	return \%info;
 }
 
-sub install_git
+sub install_git_cmds
 {
 	my ($self, $revision) = @_;
```
Comment (metan-ucw) [inline utils.pm]: The $self parameter is useless now.

File: utils.pm [hunk 5]
```diff
@@ -121,86 +121,59 @@ sub collect_sysinfo
...
+	push(@cmds, "git checkout $revision") if ($revision);
 
-	return 0;
+	return @cmds;
 }
 
-sub install_zip
+sub install_zip_cmds
 {
 	my ($self, $revision) = @_;
```
Comment (metan-ucw) [inline utils.pm]: here as well

File: utils.pm [hunk 6]
```diff
@@ -290,6 +263,49 @@ sub reboot
 	setup_ltp_run($self);
 }
 
+
+=head2 run_cmds_retry
+
+    run_cmds_retry($self, <ARRAY of commands>, [timeout => <seconds>, retries => <number>, results => <array_ref>]);
+
+Run commands sequentially. If a command failed in case of timeout, reboot the SUT.
+After reboot the sequenz is restarted from the first  command.
```
Comment (metan-ucw) [inline utils.pm]: sequence not sequenz

File: utils.pm [hunk 7]
```diff
@@ -290,6 +263,49 @@ sub reboot
 	setup_ltp_run($self);
 }
 
+
+=head2 run_cmds_retry
+
+    run_cmds_retry($self, <ARRAY of commands>, [timeout => <seconds>, retries => <number>, results => <array_ref>]);
+
+Run commands sequentially. If a command failed in case of timeout, reboot the SUT.
```
Comment (metan-ucw) [inline utils.pm]: "If command has failed because of timeout"

File: utils.pm [hunk 8]
```diff
@@ -290,6 +263,49 @@ sub reboot
...
+After reboot the sequenz is restarted from the first  command.
+The sequence stops on the first commaned which exit with none zero.
```
Comment (metan-ucw) [inline utils.pm]: "non zero"

File: utils.pm [hunk 9]
```diff
@@ -290,6 +263,49 @@ sub reboot
...
+
+The function retrieves a array of hash refs or the exitcode of the last command in scalar context:
```
Comment (metan-ucw) [inline utils.pm]: returns not retrieves

**Comment by metan-ucw** [comment]
There are some minor issues, but other than that it looks good to me.

File: backend.pm (hunk 3 above)
Comment (cfconrad) [reply backend.pm]: ok, lets do with the next!

**Comment by cfconrad** [comment]
@metan-ucw I changed now _hopeful_ all needed `run_cmd()` calls to that `utils::run_cmd_retry()` method.
The `run_cmd()` calls, related to test trigger and taint check, are not touched. From my perspective we will have here special treatment, like a **brok** result count.

**Comment by metan-ucw** [comment]
Btw it would probably make sense to add a parameter to retry some of the commands passed to tst_run_cmds() i.e. things that connect to network such as wget, git, apt-get and zypper make sense to be retried if they fail because of broken connection, but for that the command array would have to became array of tuples and the second parameter would enable retry for the particular command.

**Comment by cfconrad** [comment]
Hm I'm not sure, if I get the idea.
What I got is, you would like to pass a array out of tuple, to `backend:run_cmds()`. And if a command is marked  with retry, it will rerun the command until the exit code is `0` or limit reached. I like this idea.
What I didn't get, where `tst_run_cmds()` comes into the game.

**Comment by metan-ucw** [comment]
The tst_ is just typo, of course I meant backend:run_cmds().

**Comment by cfconrad** [comment]
ok, got it
https://github.com/metan-ucw/runltp-ng/issues/4

**Pull request merged**